### DISCUSSION TITLE
Setting constraints for Magento patches

### DISCUSSION THREAD

**Discussion by paulvandermeijs** [renovatebot/renovate#30348] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, Renovate v37.440.5

### Please tell us more about your question or problem

Hello,
I'm trying to set up Renovate to create merge requests for Magento security patches, but not normal patch releases (see https://experienceleague.adobe.com/en/docs/commerce-operations/release/planning/versioning-policy). E.g. version 2.4.6-p5 should update to 2.4.6-p6, but not 2.4.7 or higher.

I tried setting my version constraint to [`<2.4.7`](https://semver.madewithlove.com/?package=magento%2Fcommunity-edition&constraint=%3C2.4.7) which got me the following result:

```json
{
    "depType": "require",
    "depName": "magento/product-community-edition",
    "currentValue": "<2.4.7",
    "lockedVersion": "2.4.6-p3",
    "datasource": "packagist",
    "registryUrls": [
        "https://repo.magento.com/",
        "https://packagist.org"
    ],
    "updates": [
        {
            "bucket": "non-major",
            "newVersion": "2.4.7-p1",
            "newValue": "<2.4.7",
            "newMajor": 2,
            "newMinor": 4,
            "newPatch": 7,
            "updateType": "patch",
            "isRange": true,
            "isLockfileUpdate": true,
            "branchName": "renovate/magento-product-community-edition-2.x-lockfile"
        }
    ],
    "packageName": "magento/product-community-edition",
    "versioning": "composer",
    "warnings": [],
    "registryUrl": "https://repo.magento.com",
    "currentVersion": "2.4.6-p3",
    "isSingleVersion": true,
    "fixedVersion": "2.4.6-p3"
}
```

Apparently 2.4.7-p1 is seen as in range for this constraint which may be a bug.

Next I tried setting the constraint to [`~2.4.6.0`](https://semver.madewithlove.com/?package=magento%2Fcommunity-edition&constraint=~2.4.6.0) but that got me an invalid value error:

```json
{
    "depType": "require",
    "depName": "magento/product-community-edition",
    "currentValue": "~2.4.6.0",
    "lockedVersion": "2.4.6-p3",
    "datasource": "packagist",
    "registryUrls": [
        "https://repo.magento.com/",
        "https://packagist.org"
    ],
    "updates": [],
    "packageName": "magento/product-community-edition",
    "versioning": "composer",
    "warnings": [],
    "skipReason": "invalid-value",
    "currentVersion": "2.4.6-p3",
    "fixedVersion": "2.4.6-p3"
}
```

Both these constraints do get me the desired result when running `composer update`.

Is there some way I can configure Renovate to get this working?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment, marked as answer]
> Apparently 2.4.7-p1 is seen as in range for this constraint which may be a bug.

In the semver 2.0 spec anything ending in -x is a pre-release and sorts lower than the same version without a suffix. The use of -p1 to mean a stable build higher than the main version in composer is not compliant semver, and because we rely on npm's semver package we might get things wrong if we don't perform steps to massage. Constraints such as `~2.4.6.0` won't work for similar reasons.

Here's some testing I did locally:

```
    ✓ matches("1.2.7-p1", "<1.2.7") === false (1 ms)
    ✕ matches("1.2.7-p1", ">1.2.7") === true (1 ms)
    ✕ matches("1.2.7-p1", "~1.2.7") === true (1 ms)
```

Assuming you are using Renovate's `allowedVersions` feature, consider using a regex string match instead to enforce 2.4.6-p as a pattern

**Reply by paulvandermeijs** [reply]
Thanks! Using `allowedVersions` with a regular expression fixed it. Not ideal, but as long as Magento doesn't comply with the semver specification it makes sense that it will never be fully compatible with Renovate.

**Reply by rarkins** [reply]
Hopefully not never! but we'd have to rewrite the semver package for composer's needs OR find a way to hack/massage it

**Reply by HonkingGoose** [reply]
These issues seem related:

- #24993
- #28298
- #28297

**Reply by RahulGautamSingh** [reply]
@paulvandermeijs Could you create a [minimal reproduction](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md)?

**Reply by rarkins** [reply]
@RahulGautamSingh this can be sastisfied with unit tests only - see my example above of some tests which can be added
